File: kelly/문자열/문자열 압축/Solution02.java
import java.util.*;

class Solution {

    String s;

    public int solution(String s) {
        this.s = s;

        // 메인 로직
        int answer = s.length();

        for (int length = 1; length < (s.length() / 2) + 1; length++) {
            List<String> words = getWords(length);
            Deque<Item> stack = new ArrayDeque<>();
            stack.push(new Item(words.get(0), 1));

            for (int i = 1; i < words.size(); i++) {
                if (stack.peek().word.equals(words.get(i))) {
                    stack.peek().count++;
                } else {
                    stack.push(new Item(words.get(i), 1));
                }
            }

            StringBuilder sb = new StringBuilder();
            for (Item item : stack) {
                if (item.count > 1) {
                    sb.append(String.valueOf(item.count) + item.word);
                } else {
                    sb.append(item.word);
                }
            }

            answer = Math.min(answer, sb.length());
        }

        return answer;
    }

    private List<String> getWords(int length) {
        List<String> words = new ArrayList<>();

        int start = 0;
        while (start + length <= s.length()) {
            words.add(s.substring(start, start + length));
            start += length;
        }

        if (start < s.length()) {
            words.add(s.substring(start, s.length()));
        }

        return words;
    }

    class Item {
        String word;
        int count;

        public Item(String word, int count) {
            this.word = word;
            this.count = count;
        }
    }
}

File: kelly/문자열/문자열 압축/solution02.py
def solution(s):
    # 메인 로직
    if len(s) == 1:
        return 1

    answer = len(s)

    for length in range(1, (len(s) // 2) + 1):
        words = [s[i:i + length] for i in range(0, len(s), length)]
        stack = [[words[0], 1]]

        for word in words[1:]:
            if word == stack[-1][0]:
                stack[-1][1] += 1
            else:
                stack.append([word, 1])

        result = ''.join([str(cnt) + w if cnt > 1 else w for w, cnt in stack])
        answer = min(answer, len(result))

    return answer